lib/tray/tray.h

```diff
@@ -178,31 +178,53 @@ static id statusItem;
 static id statusBarButton;
 
 static id _tray_menu(struct tray_menu *m) {
-    id menu = objc_msgSend((id)objc_getClass("NSMenu"), sel_registerName("new"));
-    objc_msgSend(menu, sel_registerName("autorelease"));
-    objc_msgSend(menu, sel_registerName("setAutoenablesItems:"), false);
-
-    for (; m != NULL && m->text != NULL; m++) {
-      if (strcmp(m->text, "-") == 0) {
-        objc_msgSend(menu, sel_registerName("addItem:"),
-          objc_msgSend((id)objc_getClass("NSMenuItem"), sel_registerName("separatorItem")));
-      } else {
-        id menuItem = objc_msgSend((id)objc_getClass("NSMenuItem"), sel_registerName("alloc"));
-        objc_msgSend(menuItem, sel_registerName("autorelease"));
-        objc_msgSend(menuItem, sel_registerName("initWithTitle:action:keyEquivalent:"),
-                  objc_msgSend((id)objc_getClass("NSString"), sel_registerName("stringWithUTF8String:"), m->text),
-                  sel_registerName("menuCallback:"),
-                  objc_msgSend((id)objc_getClass("NSString"), sel_registerName("stringWithUTF8String:"), ""));
-
-        objc_msgSend(menuItem, sel_registerName("setEnabled:"), (m->disabled ? false : true));
-          objc_msgSend(menuItem, sel_registerName("setState:"), (m->checked ? 1 : 0));
-          objc_msgSend(menuItem, sel_registerName("setRepresentedObject:"),
-            objc_msgSend((id)objc_getClass("NSValue"), sel_registerName("valueWithPointer:"), m));
-
-          objc_msgSend(menu, sel_registerName("addItem:"), menuItem);
-
-          if (m->submenu != NULL) {
-            objc_msgSend(menu, sel_registerName("setSubmenu:forItem:"), _tray_menu(m->submenu), menuItem);
+  id menu = ((id (*)(id, SEL))objc_msgSend)(
+          (id)objc_getClass("NSMenu"),
+          sel_registerName("new"));
+
+  ((id (*)(id, SEL))objc_msgSend)(menu, sel_registerName("autorelease"));
+  ((id (*)(id, SEL, bool))objc_msgSend)(menu, sel_registerName("setAutoenablesItems:"), false);
```

Comment:
All of these code changes are pretty ugly.

Why is this necessary in the first place, and if it's really necessary, wouldn't it make sense to put it in a wrapper (or even a macro), so it's not as ugly as it is right now?

Reply:
Looks like `OBJC_OLD_DISPATCH_PROTOTYPES` was responsible to use the legacy variadic function.
Googling around I found https://github.com/Boscop/web-view/issues/292 and it looks like this is indeed the solution: https://github.com/Boscop/web-view/pull/294/files

Ugly. But okay then.

Reply:
@denisw You should be removing `OBJC_OLD_DISPATCH_PROTOTYPES` from https://github.com/neutralinojs/neutralinojs/blob/f09d63be9d1011f8d64a142743bd2d5f48723d40/buildzri.config.json#L82 then

Otherwise this looks good to me.

Reply:
Alight, I'll remove `OBJC_OLD_DISPATCH_PROTOTYPES` then.

And yes, I agree that it's super ugly, but I found no general way to simplify it or wrap it in a macro, and there seems no way around it - short of using Objective-C. 😓

Reply:
I assume the proper solution would be to use another high level abstraction which casts internally. But I'm not an Objective-C / Apple developer, so I wouldn't know what exists already.

Reply:
That high-level abstraction is exactly what Objective-C is, I'm afraid. There seems to be no way around the manual casts when sending messages from C, and because every method signature requires a different cast, it's hard to abstract.

Reply:
`OBJC_OLD_DISPATCH_PROTOTYPES` is now removed. The build seems to still succeed.